File: scripts/quad_simulator.py
#!/usr/bin/env python_cyecca
import sys
import time
import numpy as np

import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from rclpy.time import Time
from rclpy.lifecycle import LifecycleNode
import rclpy.clock

from rosgraph_msgs.msg import Clock
from sensor_msgs.msg import Imu, NavSatFix, MagneticField
from actuator_msgs.msg import Actuators

import cyecca
import casadi as ca


class QuadSimulator(LifecycleNode):

    def __init__(self, **kwargs):
        super().__init__(**kwargs)

        # parameters
        self.declare_parameter("dt", 0.005)
        self.declare_parameter("speed", 10)
        self.declare_parameter("monitor_steps", 1000)
        self.declare_parameter("dt_imu", 0.005)
        self.declare_parameter("dt_gnss", 0.1)
        self.declare_parameter("dt_mag", 0.02)
        self.set_parameters([
            Parameter("use_sim_time", Parameter.Type.BOOL, True),
        ])

        # clock
        self.pub_clock = self.create_publisher(Clock, "/clock", 10)
        self.dt = self.get_parameter("dt").value
        self.speed = self.get_parameter("speed").value
        self.monitor_steps = self.get_parameter("monitor_steps").value
        self.system_clock = rclpy.clock.Clock()
        self.create_timer(
            timer_period_sec=self.dt/self.speed,
            callback=self.update_clock,
            clock=self.system_clock)
        self.i = 0
        self.system_time_prev = self.system_clock.now()
        self.msg_clock = Clock()

        # imu
        self.pub_imu = self.create_publisher(Imu, "/imu", 10)
        self.dt_imu = self.get_parameter("dt_imu").value
        self.create_timer(
            timer_period_sec=self.dt_imu,
            callback=self.update_imu,
            clock=self.get_clock())
        self.msg_imu = Imu()
        self.msg_imu.header.frame_id = 'base_link'

        # gnss
        self.pub_gnss = self.create_publisher(NavSatFix, "/gnss", 10)
        self.dt_gnss = self.get_parameter("dt_gnss").value
        self.create_timer(
            timer_period_sec=self.dt_gnss,
            callback=self.update_gnss,
            clock=self.get_clock())
        self.msg_gnss = NavSatFix()
        self.msg_gnss.header.frame_id = 'base_link'
        self.msg_gnss.position_covariance = np.diag([1.0, 1.0, 1.0]).reshape(-1)
        self.msg_gnss.position_covariance_type = 1 # diagonal known

        # mag
        self.pub_mag = self.create_publisher(MagneticField, "/mag", 10)
        self.dt_mag = self.get_parameter("dt_mag").value
        self.create_timer(
            timer_period_sec=self.dt_mag,
            callback=self.update_mag,
            clock=self.get_clock())
        self.msg_mag = MagneticField()
        self.msg_mag.header.frame_id = 'base_link'
        self.msg_mag.magnetic_field_covariance = np.diag([1.0, 1.0, 1.0]).reshape(-1)

        # control
        self.motors = [0, 0, 0, 0]
        self.sub_actuators = self.create_subscription(
                Actuators, "/actuators", self.callback_actuators, 10)


    def update_clock(self):
        t = self.i * self.dt
        seconds = int(t)
        nanoseconds = int(1e9*(t - seconds))
        self.msg_clock.clock = Time(
            seconds=seconds,
            nanoseconds=nanoseconds,
            clock_type=rclpy.clock.ClockType.ROS_TIME).to_msg()
        self.pub_clock.publish(self.msg_clock)

        if self.i > 0 and self.i % self.monitor_steps == 0: 
            now = self.system_clock.now()
            sim_rate = self.monitor_steps*1e9*self.dt/(now - self.system_time_prev).nanoseconds
            self.system_time_prev = now
            self.get_logger().info(f"sim rate: {sim_rate:f}", throttle_duration_sec=1.0)

        # increment time for next loop
        self.i += 1

        # update dynamics
        self.update_dynamcis()

    def update_dynamcis(self):
        pass

    def update_imu(self):
        self.msg_imu.header.stamp = self.get_clock().now().to_msg()
        self.msg_imu.angular_velocity.x = 1.0
        self.msg_imu.angular_velocity.y = 2.0
        self.msg_imu.angular_velocity.z = 3.0
        self.pub_imu.publish(self.msg_imu)

    def update_gnss(self):
        self.msg_gnss.header.stamp = self.get_clock().now().to_msg()
        self.msg_gnss.altitude = 0.0
        self.msg_gnss.latitude = 1.0
        self.msg_gnss.longitude = 2.0
        self.pub_gnss.publish(self.msg_gnss)

    def update_mag(self):
        self.msg_mag.header.stamp = self.get_clock().now().to_msg()
        self.msg_mag.magnetic_field.x = 1.0
        self.msg_mag.magnetic_field.y = 2.0
        self.msg_mag.magnetic_field.z = 2.0
        self.pub_mag.publish(self.msg_mag)

    def callback_actuators(self, msg: Actuators):
        for i in range(4):
            self.motors[i] = msg.velocity[i]


def main(args=None):
    try:
        rclpy.init(args=args)
        simulator = QuadSimulator(node_name="quad_simulator")
        rclpy.spin(simulator)
    except KeyboardInterrupt:
        pass
    simulator.destroy_node()
    rclpy.try_shutdown()

if __name__ == "__main__":
    main()

File: src/quad_simulator.cpp
#include <chrono>

#include <actuator_msgs/msg/actuators.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_lifecycle/lifecycle_node.hpp>
#include <rosgraph_msgs/msg/clock.hpp>
#include <sensor_msgs/msg/imu.hpp>
#include <sensor_msgs/msg/magnetic_field.hpp>
#include <sensor_msgs/msg/nav_sat_fix.hpp>

using namespace std::placeholders;

class QuadSimulator : public rclcpp_lifecycle::LifecycleNode {
public:
    QuadSimulator();

private:
    // private methods
    void m_timer_clock_callback();
    void m_update_dynamics();
    void m_timer_imu_callback();
    void m_timer_gnss_callback();
    void m_timer_mag_callback();
    void m_actuators_callback(const actuator_msgs::msg::Actuators& msg);

    // clock
    double m_dt;
    double m_speed;
    int m_i { 0 };
    rclcpp::TimerBase::SharedPtr m_timer_clock;
    rclcpp::Publisher<rosgraph_msgs::msg::Clock>::SharedPtr m_pub_clock;
    rclcpp::Time m_system_time_prev;
    rosgraph_msgs::msg::Clock m_msg_clock;
    rclcpp::Clock::SharedPtr m_system_clock;

    // imu
    rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr m_pub_imu;
    double m_dt_imu;
    rclcpp::TimerBase::SharedPtr m_timer_imu;
    sensor_msgs::msg::Imu m_msg_imu;

    // gnss
    rclcpp::Publisher<sensor_msgs::msg::NavSatFix>::SharedPtr m_pub_gnss;
    double m_dt_gnss;
    rclcpp::TimerBase::SharedPtr m_timer_gnss;
    sensor_msgs::msg::NavSatFix m_msg_gnss;

    // mag
    rclcpp::Publisher<sensor_msgs::msg::MagneticField>::SharedPtr m_pub_mag;
    double m_dt_mag;
    rclcpp::TimerBase::SharedPtr m_timer_mag;
    sensor_msgs::msg::MagneticField m_msg_mag;

    // control
    std::array<double, 4> m_motor_angular_velocity;
    rclcpp::Subscription<actuator_msgs::msg::Actuators>::SharedPtr m_sub_actuators;
};

QuadSimulator::QuadSimulator()
    : LifecycleNode("quad_simulator")
{
    // initialize system clock
    m_system_clock = std::make_shared<rclcpp::Clock>(RCL_SYSTEM_TIME);

    // parameters
    this->declare_parameter("dt", 0.005);
    this->declare_parameter("speed", 1.0);
    this->declare_parameter("dt_imu", 0.005);
    this->declare_parameter("dt_gnss", 0.1);
    this->declare_parameter("dt_mag", 0.02);
    this->set_parameter(rclcpp::Parameter("use_sim_time", true));

    // clock
    m_pub_clock = this->create_publisher<rosgraph_msgs::msg::Clock>("clock", 10);
    m_dt = this->get_parameter("dt").as_double();
    m_speed = this->get_parameter("speed").as_double();
    int64_t nanos_sleep = 1e9 * m_dt / m_speed;
    RCLCPP_INFO(this->get_logger(), "requested sim speed: %10.4f Hz", m_speed);
    RCLCPP_INFO(this->get_logger(), "sim sleep period: %ld ns", nanos_sleep);
    m_timer_clock = rclcpp::create_timer(
        this,
        m_system_clock,
        std::chrono::nanoseconds(nanos_sleep),
        std::bind(&QuadSimulator::m_timer_clock_callback, this));
    m_system_time_prev = m_system_clock->now();

    // imu
    m_pub_imu = this->create_publisher<sensor_msgs::msg::Imu>("imu", 10);
    m_dt_imu = this->get_parameter("dt_imu").as_double();
    m_timer_imu = rclcpp::create_timer(
        this,
        this->get_clock(),
        std::chrono::nanoseconds(int64_t(1e9 * m_dt_imu)),
        std::bind(&QuadSimulator::m_timer_imu_callback, this));
    m_msg_imu.header.frame_id = "base_link";
    m_msg_imu.angular_velocity_covariance = {
        1, 0, 0,
        0, 1, 0,
        0, 0, 1
    };
    m_msg_imu.linear_acceleration_covariance = {
        1, 0, 0,
        0, 1, 0,
        0, 0, 1
    };

    // gnss
    m_pub_gnss = this->create_publisher<sensor_msgs::msg::NavSatFix>("gnss", 10);
    m_dt_gnss = this->get_parameter("dt_gnss").as_double();
    m_timer_gnss = rclcpp::create_timer(
        this,
        this->get_clock(),
        std::chrono::nanoseconds(int64_t(1e9 * m_dt_gnss)),
        std::bind(&QuadSimulator::m_timer_gnss_callback, this));
    m_msg_gnss.header.frame_id = "base_link";
    m_msg_gnss.position_covariance = {
        1, 0, 0,
        0, 1, 0,
        0, 0, 1
    };
    m_msg_gnss.position_covariance_type = 1; // diagonal known

    // mag
    m_pub_mag = this->create_publisher<sensor_msgs::msg::MagneticField>("mag", 10);
    m_dt_mag = this->get_parameter("dt_mag").as_double();
    m_timer_mag = rclcpp::create_timer(
        this,
        this->get_clock(),
        std::chrono::nanoseconds(int64_t(1e9 * m_dt_mag)),
        std::bind(&QuadSimulator::m_timer_mag_callback, this));
    m_msg_mag.header.frame_id = "base_link";
    m_msg_mag.magnetic_field_covariance = {
        1, 0, 0,
        0, 1, 0,
        0, 0, 1
    };

    // control
    m_sub_actuators = this->create_subscription<actuator_msgs::msg::Actuators>(
        "actuators", 10, std::bind(&QuadSimulator::m_actuators_callback, this, _1));
}

void QuadSimulator::m_timer_clock_callback()
{
    double t = m_i * m_dt;
    double sec = int(t);
    double nanosec = int(1e9 * (t - sec));
    m_msg_clock.clock.sec = sec;
    m_msg_clock.clock.nanosec = nanosec;
    m_pub_clock->publish(m_msg_clock);

    auto now = m_system_clock->now();
    double sim_rate = 1e9 * m_dt / (now - m_system_time_prev).nanoseconds();
    m_system_time_prev = now;
    RCLCPP_INFO_THROTTLE(
        this->get_logger(), *m_system_clock,
        1000, "sim speed:%10.3f Hz", sim_rate);

    // increment time for next loop
    m_i += 1;

    // update dynamics
    m_update_dynamics();
}

void QuadSimulator::m_update_dynamics()
{
}

void QuadSimulator::m_timer_imu_callback()
{
    m_msg_imu.header.stamp = this->get_clock()->now();
    m_msg_imu.angular_velocity.x = 1.0;
    m_msg_imu.angular_velocity.y = 2.0;
    m_msg_imu.angular_velocity.z = 3.0;
    m_pub_imu->publish(m_msg_imu);
}

void QuadSimulator::m_timer_gnss_callback()
{
    m_msg_gnss.header.stamp = this->get_clock()->now();
    m_msg_gnss.altitude = 1.0;
    m_msg_gnss.latitude = 2.0;
    m_msg_gnss.longitude = 3.0;
    m_pub_gnss->publish(m_msg_gnss);
}

void QuadSimulator::m_timer_mag_callback()
{
    m_msg_mag.header.stamp = this->get_clock()->now();
    m_msg_mag.magnetic_field.x = 1.0;
    m_msg_mag.magnetic_field.y = 2.0;
    m_msg_mag.magnetic_field.z = 3.0;
    m_pub_mag->publish(m_msg_mag);
}

void QuadSimulator::m_actuators_callback(const actuator_msgs::msg::Actuators& msg)
{
    for (int i = 0; i < 4; i++) {
        m_motor_angular_velocity[i] = msg.velocity[i];
    }
}

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::executors::SingleThreadedExecutor exe;
    auto quad_sim = std::make_shared<QuadSimulator>();
    exe.add_node(quad_sim->get_node_base_interface());
    exe.spin();
    rclcpp::shutdown();
    return 0;
}

// vi: ts=4 sw=4 et
